rsocket-core/src/main/java/io/rsocket/core/RSocketRequester.java

```diff
@@ -560,46 +560,58 @@ private void handleStreamZero(FrameType type, ByteBuf frame) {
 
   private void handleFrame(int streamId, FrameType type, ByteBuf frame) {
     Subscriber<Payload> receiver = receivers.get(streamId);
-    if (receiver == null) {
-      handleMissingResponseProcessor(streamId, type, frame);
-    } else {
-      switch (type) {
-        case ERROR:
-          receiver.onError(Exceptions.from(streamId, frame));
-          receivers.remove(streamId);
-          break;
-        case NEXT_COMPLETE:
-          receiver.onNext(payloadDecoder.apply(frame));
-          receiver.onComplete();
-          break;
-        case CANCEL:
-          {
-            Subscription sender = senders.remove(streamId);
-            if (sender != null) {
-              sender.cancel();
-            }
-            break;
+    switch (type) {
+      case NEXT:
+        if (receiver == null) {
+          handleMissingResponseProcessor(streamId, type, frame);
+          return;
+        }
+        receiver.onNext(payloadDecoder.apply(frame));
+        break;
+      case NEXT_COMPLETE:
+        if (receiver == null) {
+          handleMissingResponseProcessor(streamId, type, frame);
+          return;
+        }
+        receiver.onNext(payloadDecoder.apply(frame));
+        receiver.onComplete();
+        break;
+      case COMPLETE:
+        if (receiver == null) {
+          handleMissingResponseProcessor(streamId, type, frame);
+          return;
+        }
+        receiver.onComplete();
+        receivers.remove(streamId);
+        break;
+      case ERROR:
+        if (receiver == null) {
+          handleMissingResponseProcessor(streamId, type, frame);
+          return;
+        }
+        receiver.onError(Exceptions.from(streamId, frame));
+        receivers.remove(streamId);
+        break;
+      case CANCEL:
+        {
+          Subscription sender = senders.remove(streamId);
+          if (sender != null) {
+            sender.cancel();
           }
-        case NEXT:
-          receiver.onNext(payloadDecoder.apply(frame));
           break;
-        case REQUEST_N:
-          {
-            Subscription sender = senders.get(streamId);
-            if (sender != null) {
-              int n = RequestNFrameFlyweight.requestN(frame);
-              sender.request(n >= Integer.MAX_VALUE ? Long.MAX_VALUE : n);
-            }
-            break;
+        }
+      case REQUEST_N:
+        {
+          Subscription sender = senders.get(streamId);
+          if (sender != null) {
+            long n = RequestNFrameFlyweight.requestN(frame);
```

Comment:
is that the bug you're mentioning in the PR body? ie if the `RequestNFrameFlyweight.requestN` already used to return a `long` that was cast into an `int`, meaning that any request in the range `Integer.MAX_VALUE..Long.MAX_VALUE` was previously propagated as `Long.MAX_VALUE` ? 

In any case, I would suggest creating a dedicated issue describing this bug for reference (this PR can fix both the partial close and the request bug, but at least the later becomes searchable in issues, which is likely needed)

Reply:
That was pulled from a different PR acidentially

Reply:
The Bug if that RequestChannel on the requester side has 2 maps to track the state:
 - `receivers` which is there to receive OnNext/OnError/OnComplete signals 
 - `senders` -  requestChannel specific map to hold a subscription and receive the request/cancel signals from the remote.

Originally, we had that -> https://github.com/rsocket/rsocket-java/blob/develop/rsocket-core/src/main/java/io/rsocket/core/RSocketRequester.java#L562

```
Subscriber<Payload> receiver = receivers.get(streamId);
    if (receiver == null) {
      handleMissingResponseProcessor(streamId, type, frame);
    }
```

which means that if there is no receiver - it is impossible to handle any frames.

However, due to the spec, there is half-closed state when the responder may send incomplete, but the requester may still continue its stream and send onXXX signals along with being able to receive demand (requestN) and cancel frames.

Reply:
yeah, I figured this was a separate issue. ok then, separate PR is good 👍